react_juce/core/ImageView.cpp

```diff
@@ -0,0 +1,228 @@
+                {
+                    auto drawableImg = std::make_unique<juce::DrawableImage>();
+                    drawableImg->setImage(loadImageFromFileURL(sourceURL));
+                    drawable = std::move(drawableImg);
+                }
+                else
+                {
+                    const juce::String errorString = "Unsupported image URL: " + source;
+                    throw std::logic_error(errorString.toStdString());
+                }
+            }
+            else if (source.startsWith("data:image/")) // juce::URL does not currently handle Data URLs
+            {
+                auto drawableImg = std::make_unique<juce::DrawableImage>();
+                drawableImg->setImage(loadImageFromDataURL(source));
+                drawable = std::move(drawableImg);
+            }
+            else // If not a URL treat source prop as inline SVG/Image data
+            {
+                drawable = std::unique_ptr<juce::Drawable>(
+                    juce::Drawable::createFromImageData(
+                        source.toRawUTF8(),
+                        source.getNumBytesAsUTF8()
+                    )
+                );
+            }
+        }
+    }
+
+    //==============================================================================
+    void ImageView::paint(juce::Graphics& g)
+    {
+        View::paint(g);
+
+        if (drawable == nullptr) return;
+
+        const float opacity = props.getWithDefault(opacityProp, 1.0f);
+
+        // Without a specified placement, we just draw the drawable.
+        if (!props.contains(placementProp))
+            return drawable->draw(g, opacity);
+
+        // Otherwise we map placement strings to the appropriate flags
+        const int existingFlags = props[placementProp];
+        const juce::RectanglePlacement placement(existingFlags);
+
+        drawable->drawWithin(g, getLocalBounds().toFloat(), placement, opacity);
+    }
+
+    //==============================================================================
+    void ImageView::parentHierarchyChanged()
+    {
+        if (shouldDownloadImage)
+        {
+            downloadImageAsync(props[sourceProp].toString());
+        }
+    }
+
+    //==============================================================================
+    void ImageView::downloadImageAsync(const juce::String& source)
+    {
+        // Allow only one download at the same time for a given ImageView.
+        if (downloading)
+            return;
+
+        if (auto* appRoot = findParentComponentOfClass<ReactApplicationRoot>())
+        {
+            appRoot->getThreadPool().addJob([this, source] ()
+            {
+                downloading = true;
+                juce::MemoryBlock mb;
+
+                if (!juce::URL(source).readEntireBinaryStream(mb)) {
+                    downloading = false;
+                    shouldDownloadImage = true;
+                    throw std::runtime_error("Failed to fetch the image!");
```

Comment:
Instead of throwing here, let's enqueue a different call to `juce::MessageManager::callAsync` that basically just invokes `props["onError"]` if the user has supplied such a prop. Else maybe we just log to the console

Reply:
Good one! So just to be sure, it works like that?
```cpp
if (props.contains(onErrorProp) && props[onErrorProp].isMethod())
    std::invoke( props[onErrorProp].getNativeFunction() );
```

Reply:
Yup, here's an example: https://github.com/nick-thompson/blueprint/blob/master/react_juce/core/CanvasView.h#L718